### PR TITLE
Merging `thesis_gemein_2022` into `TrixiAtmo.jl`

### PR DESCRIPTION
**Pull request by Arpit-Babbar**
This PR started as an explanation to https://github.com/trixi-framework/TrixiAtmo.jl/issues/1. Now, it adds all the elixirs with respective CI tests from Lucas' thesis.

TODO 
- [x] Add tests for elixirs
- [ ] Resolve repetition from Trixi in the process of adding tests. That is, this file https://github.com/trixi-framework/TrixiAtmo.jl/blob/ab_lucas/test/test_trixiatmo.jl is copied from Trixi.jl.
- [x] Are the 1920 allocations expected? (They didn't show up in CI, so we are good)
- [x] Add tests for the remaining 5/7 elixirs

### CODE REVIEW

**Comment by tristanmontoya** [comment]
This is a good start for TrixiAtmo.jl, and should be quite useful. I will have some time to look into the specifics more in the next week, but my main comment is that we should perhaps set up testing/CI (similarly to TrixiShallowWater.jl) before merging this to main.

File: examples/elixir_moist_euler_dry_bubble.jl [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    r = sqrt((x[1] - xc)^2 + (x[2] - zc)^2)
+    rc = 2000.0
+    θ_ref = 300.0
+    Δθ = 0.0
+
+    if r <= rc
+        Δθ = 2 * cospi(0.5 * r / rc)^2
+    end
+
+    #Perturbed state:
```
Comment (Arpit-Babbar) [inline examples/elixir_moist_euler_dry_bubble.jl:28]:
```suggestion
    # Perturbed state:
```

File: examples/elixir_moist_euler_dry_bubble.jl [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    θ = θ_ref + Δθ # potential temperature
+    # π_exner = 1 - g / (c_pd * θ) * x[2] # exner pressure
+    # rho = p_0 / (R_d * θ) * (π_exner)^(c_vd / R_d) # density
+
+    # calculate background pressure with assumption hydrostatic and neutral
+    p = p_0 * (1 - kappa * g * x[2] / (R_d * θ_ref))^(c_pd / R_d)
+
+    #calculate rho and T with p and theta (now perturbed) rho = p / R_d T, T = θ / π
```
Comment (Arpit-Babbar) [inline examples/elixir_moist_euler_dry_bubble.jl:36]:
```suggestion
    # calculate rho and T with p and theta (now perturbed) rho = p / R_d T, T = θ / π
```

File: examples/elixir_moist_euler_dry_bubble.jl [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    rho = p / ((p / p_0)^kappa * R_d * θ)
+    T = p / (R_d * rho)
+
+    v1 = 20.0
+    #v1 = 0.0
```
Comment (Arpit-Babbar) [inline examples/elixir_moist_euler_dry_bubble.jl:41]:
```suggestion
    # v1 = 0.0
```

File: src/equations/compressible_moist_euler_2d.jl [hunk 4]
```diff
@@ -0,0 +1,1029 @@
...
+        mu = ((1 - qv - ql) * c_vd + qv * c_vv + ql * c_pl)
+        xi = ((1 - qv - ql) * R_d + qv * R_v)
+
+        T = (rho - 1) / mu + 10 / c_vd + 40
+        dT = rho_x / c_vd
+        E = (mu * T + qv * L_00 + 1)
+        dE = E * rho_x + rho * mu * dT
+        dp = xi * (T * rho_x + rho * dT)
+
+        #Calculate Error in Sources with exact solution and u
```
Comment (Arpit-Babbar) [inline src/equations/compressible_moist_euler_2d.jl:312]:
```suggestion
        # Calculate Error in Sources with exact solution and u
```

File: src/equations/compressible_moist_euler_2d.jl [hunk 5]
```diff
@@ -0,0 +1,1029 @@
...
+        u_exact = SVector(rho, rho, rho, rho * E, rho * qv, rho * ql)
+
+        du1, du2, du3, du4, du5, du6 = (source_terms_moist_bubble(u, x, t, equations) -
+                                        source_terms_moist_bubble(u_exact, x, t, equations))
+        #du1, du2, du3, du4, du5, du6 = zeros(Float64, 6)
```
Comment (Arpit-Babbar) [inline src/equations/compressible_moist_euler_2d.jl:317]:
```suggestion
        # du1, du2, du3, du4, du5, du6 = zeros(Float64, 6)
```

File: src/equations/compressible_moist_euler_2d.jl [hunk 6]
```diff
@@ -0,0 +1,1029 @@
...
+        rho, rho_v1, rho_v2, rho_e, rho_qv, rho_ql = u
+        tmp = rho_v2
+
+        return SVector(zero(eltype(u)), zero(eltype(u)),
+                       -g * rho, -g * tmp,
+                       zero(eltype(u)), zero(eltype(u)))
+    end
+
+    # Raylight damping sponge source term form A. Sridhar et al.,
+    #Large-eddy simulations with ClimateMachine: a new open-sourcecode for
```
Comment (Arpit-Babbar) [inline src/equations/compressible_moist_euler_2d.jl:354]:
```suggestion
    # Large-eddy simulations with ClimateMachine: a new open-sourcecode for
```

File: src/equations/compressible_moist_euler_2d.jl [hunk 7]
```diff
@@ -0,0 +1,1029 @@
...
+    #atmospheric simulations on GPUs and CPUs, 2 Oct 2021, doi: 10.5194/gmd-15-6259-2022,
```
Comment (Arpit-Babbar) [inline src/equations/compressible_moist_euler_2d.jl:355]:
```suggestion
    # atmospheric simulations on GPUs and CPUs, 2 Oct 2021, doi: 10.5194/gmd-15-6259-2022,
```

File: src/equations/compressible_moist_euler_2d.jl [hunk 8]
```diff
@@ -0,0 +1,1029 @@
...
+    #https://arxiv.org/abs/2110.00853 [physics.ao-ph] .
```
Comment (Arpit-Babbar) [inline src/equations/compressible_moist_euler_2d.jl:356]:
```suggestion
    # https://arxiv.org/abs/2110.00853 [physics.ao-ph] .
```

File: src/equations/compressible_moist_euler_2d.jl [hunk 9]
```diff
@@ -0,0 +1,1029 @@
...
+
+        # relaxed background velocity
+        vr1, vr2 = (10.0, 0.0)
+        # damping threshold
+        z_s = 9000.0
+        # boundary top
+        z_top = 16000.0
+        # positive even power with default value 2
+        gamma = 2.0
+        #relaxation coefficient > 0
```
Comment (Arpit-Babbar) [inline src/equations/compressible_moist_euler_2d.jl:372]:
```suggestion
        # relaxation coefficient > 0
```

File: src/equations/compressible_moist_euler_2d.jl [hunk 10]
```diff
@@ -0,0 +1,1029 @@
...
+        @unpack R_d, R_v, c_pd, c_pv, c_pl, L_00 = equations
+        rho, rho_v1, rho_v2, rho_E, rho_qv, rho_ql = u
+
+        v1 = rho_v1 / rho
+        v2 = rho_v2 / rho
+        p, T = get_current_condition(u, equations)
+        v_square = v1^2 + v2^2
+        rho_qd = rho - rho_qv - rho_ql
+
+        # Work around if an individual density is zero
```
Comment (Arpit-Babbar) [inline src/equations/compressible_moist_euler_2d.jl:531]:
```suggestion
        # Workaround if an individual density is zero
```

File: src/equations/compressible_moist_euler_2d.jl [hunk 11]
```diff
@@ -0,0 +1,1029 @@
...
+        T_C = T - 273.15
+        p_vs = 611.2 * exp(17.62 * T_C / (243.12 + T_C))
+        H = p_v / p_vs
+        r_v = rho_qv / rho_d
+        r_l = rho_ql / rho_d
+        r_t = r_v + r_l
+        L_v = L_00 + (c_pv - c_pl) * T
+        c_p = c_pd + r_t * c_pl
+
+        #equivalentpotential temperature
```
Comment (Arpit-Babbar) [inline src/equations/compressible_moist_euler_2d.jl:829]:
```suggestion
        # equivalent potential temperature
```

File: src/equations/compressible_moist_euler_2d.jl [hunk 12]
```diff
@@ -0,0 +1,1029 @@
...
```
Comment (Arpit-Babbar) [inline src/equations/compressible_moist_euler_2d.jl:855]:
```suggestion
        # equivalent potential temperature
```

**Review by Arpit-Babbar — COMMENTED** [review]
Minor fixes

**Comment by benegee** [comment]
> * Resolve repetition from Trixi in the process of adding tests. That is, this file https://github.com/trixi-framework/TrixiAtmo.jl/blob/ab_lucas/test/test_trixiatmo.jl is copied from Trixi.jl.

At first I though that we could maybe move some parts to TrixiBase.jl, but then I realized that these macro e.g. explicitly include the `Trixi`. Maybe we should there add our own macro as need be.

**Comment by benegee** [comment]
> * Are the 1920 allocations expected?

Which test case was it?

File: test/test_2d_moist_euler.jl [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+                        tspan=(0.0, 0.1))
+    # Ensure that we do not have excessive memory allocations
+    # (e.g., from type instabilities)
+    let
+        t = sol.t[end]
+        u_ode = sol.u[end]
+        du_ode = similar(u_ode)
+        analysis_callback(sol)
+        @test (@allocated TrixiAtmo.Trixi.rhs!(du_ode, u_ode, semi, t)) < 2000
+        # TODO - It performs 1920 allocations. Is that okay? All tests in Trixi take 1000.
```
Comment (Arpit-Babbar) [inline test/test_2d_moist_euler.jl:71]: Both of these tests take 1920 allocations @benegee 

The elixirs are https://github.com/trixi-framework/TrixiAtmo.jl/blob/ab_lucas/examples/elixir_moist_euler_EC_bubble.jl and https://github.com/trixi-framework/TrixiAtmo.jl/blob/ab_lucas/examples/elixir_moist_euler_dry_bubble.jl.

Comment (benegee) [reply test/test_2d_moist_euler.jl:71]: Thanks!

In fact, it works for me and I see 0 allocations for both tests. We can also try it here using the CI tests.

Comment (benegee) [reply test/test_2d_moist_euler.jl:71]: I merge current main and the tests pass.
Note that the `test_trixi_include` macro disables some test when it detects coverage testing. We should adapt this to our needs. 
Anyway, allocations seem to be alright?

Comment (Arpit-Babbar) [reply test/test_2d_moist_euler.jl]: Yeah, I see that the allocations do not appear in CI. Thanks for testing it!

**Review by tristanmontoya — DISMISSED** [review]
It looks good to me, although I am wondering if either of you (@Arpit-Babbar or @benegee) know the reason for the warnings copied below? My tests have always been very simple so I've never run into this. 

`WARNING: replacing module TrixiAtmoTestModule.`
`WARNING: could not import TestExamples2DMoistEuler.TRIXI_EXAMPLES_DIR into TrixiAtmoTestModule`

**Review by tristanmontoya — COMMENTED** [review]
I think it's good to merge if we're ok with the warnings I mentioned above.

**Comment by ranocha** [comment]
> It looks good to me, although I am wondering if either of you (@Arpit-Babbar or @benegee) know the reason for the warnings copied below? My tests have always been very simple so I've never run into this.
> 
> `WARNING: replacing module TrixiAtmoTestModule.` `WARNING: could not import TestExamples2DMoistEuler.TRIXI_EXAMPLES_DIR into TrixiAtmoTestModule`

You define a module with the same name every time in

https://github.com/trixi-framework/TrixiAtmo.jl/blob/1c7efd9be9a80019df8d74bb765ffdccd6a90ba1/test/test_trixiatmo.jl#L28

If you want to get rid of that warning, you could `gensym` an appropriate name - but I don't think that's necessary right now.

**Review by tristanmontoya — APPROVED** [review]

**Pull request merged**